### PR TITLE
Tooling: migrated dt-header to ESLint, with a script

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Addresses https://github.com/microsoft/DefinitelyTyped-tools/pull/514#pullrequestreview-1074873002 by introducing a script that, given a TSLint rule name and ESLint rule name (which defaults to the TSLint rule name):

1. Finds all types packages that have a `tslint.json` disabling the rule
2. Deletes the rule from the `tslint.json` -- or removes the `tslint.json` altogether if it's now empty
3. Adds the rule disable to `.eslintrc.json` -- creating the `.eslintrc.json` file if it didn't yet exists

The result of running `node scripts/tslint-rule-to-eslint dt-header` is checked in.

Note that the changes from https://github.com/microsoft/DefinitelyTyped-tools/pull/514 need to be available for the `dt-header` in the root-level `.eslintrc.json`.

### CODE REVIEW

**Comment by sandersn** [comment]
I haven't read the PR yet, but regarding (2), tslint.json is always required, but should ideally have no `rules` object when there are no overrides. However, if there is an empty `rules` object, I *think* that's equivalent.

File: scripts/tslint-rule-to-eslint.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        console.log(`Converting ${typeName}...`);
+
+        delete tslintData.rules[tslintRuleName];
+        if (Object.keys(tslintData.rules).length === 0) {
+            console.log(`\t${tslintFilePath} has no remaining rules; deleting.`);
+            await fs.rm(tslintFilePath);
+        } else {
+            console.log(`\t${tslintFilePath} has remaining rules; modifying.`);
+            await writeFileFormatted(tslintFilePath, tslintData);
+        }
```
Comment (sandersn) [inline scripts/tslint-rule-to-eslint.js:50]:
```suggestion
            delete tslintFilePath.rules
        } 
        await writeFileFormatted(tslintFilePath, tslintData);
```

**Review by sandersn — COMMENTED** [review]
I suggested the tslint change I mentioned earlier.

**Comment by sandersn** [comment]
Looks good so far; I'm going to close/reopen to retrigger CI.

**Comment by sandersn** [comment]
Still quite a few missing tslint.jsons. I'm running locally and will copy in default ones where they're missing.

**Comment by sandersn** [comment]
The script doesn't cover types/some-package/v*N* subdirectories, which means that it misses 5 entries. I'm going to skip those for now.

**Comment by sandersn** [comment]
I manually fixed the missing v* entries but didn't fix the script since I'm doing this largely as a background task and don't want to stop to write code.

**Comment by sandersn** [comment]
Ugh, I found 68 more v* packages that turn off dt-header, which causes tslint to error? warn?, but they don't fail on the eslint version. The fixup script definitely needs to support v* and ts* subdirectories.

File: scripts/tslint-rule-to-eslint.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import { parse } from 'comment-json';
+import { promises as fs } from 'fs';
+import { format } from 'prettier';
+import * as path from 'path';
+
+const emptyObject = {};
+
+const parseAndReadFileContents = async filePath => {
+    try {
+        return parse((await fs.readFile(filePath)).toString());
```
Comment (peterblazejewicz) [inline scripts/tslint-rule-to-eslint.js:10]:
```ts
        const configFile =  await prettier.resolveConfigFile(filePath);
        const config  =  await prettier.resolveConfig(configFile);
        if (config) {
            config.editorconfig  = true;
        }
        return config;
```
IMMO that would use existing formatting setup (EditorConfig/Pretier) and remove the need for new dependency

File: package.json [hunk 3]
```diff
@@ -31,6 +31,7 @@
         "@definitelytyped/utils": "latest",
         "@octokit/core": "^3.5.1",
         "@octokit/rest": "^16.0.0",
+        "comment-json": "^4.2.3",
```
Comment (peterblazejewicz) [inline package.json:34]: IMO this dependency can be avoided IMO when using prettier own config lookup methods, see other comment

Comment (JoshuaKGoldberg) [reply package.json:34]: Ah, but `tslint.json` files can have comments too.

File: scripts/tslint-rule-to-eslint.js (hunk 2 above)
Comment (JoshuaKGoldberg) [reply scripts/tslint-rule-to-eslint.js:10]: Nice, +1! Thanks!

File: package.json (hunk 3 above)
Comment (peterblazejewicz) [reply package.json:34]: oh, you're absolutely correct (no surprise here regarding topic, btw). Thanks!

File: scripts/tslint-rule-to-eslint.js (hunk 2 above)
Comment (peterblazejewicz) [reply scripts/tslint-rule-to-eslint.js:10]: Just review in context of JSON5, as you mentioned in other place. Prettier supports only with 'json5' extension. Thanks!

**Comment by sandersn** [comment]
I looked at the CI failures. Nothing left from the dt-header migration; I saw most of these this morning from the overnight run. I'm going to merge this and @JoshuaKGoldberg you or I can follow up with fixes to the script as well as running it on the remaining 68 v* subdirectories.

**Comment by JoshuaKGoldberg** [comment]
Oh! @sandersn I _just_ pushed up a fix to the branch. Amusing. I'll send a PR.

**Comment by sandersn** [comment]
sorry =(

**Comment by forivall** [comment]
For the eslint plugin in vscode, I had to set the following options in `.vscode/settings.json`:

```json
{
  "eslint.options": {
    "rulePaths": ["node_modules/@definitelytyped/dtslint/dist/rules"]
  }
}
```

It might be worthwhile to include that config in the repo, or, make an eslint plugin for the custom rules, for ease of editor and other tooling integration.

**Comment by jakebailey** [comment]
That's a good idea, though I'm not quite sure what to do about people who just open a single directory in the repo (versus the entire thing); the vscode setting won't apply. Maybe a plugin with an eslintrc at the top level is the best choice. IIRC you can open a subdirectory of a repo, and the ESLint extension will walk upward.

I don't think we had editor support before for dtslint rules, so I'm guessing that switching to ESLint without it is not a regression, thankfully.

**Comment by forivall** [comment]
Yeah, I'm aware of that, as I am one of those people who just opens a single directory. Which is why I suggested the plugin.

The TSLint DTSLint rules actually did work with the TSLint plugin, so it's only a regression for an unsupported/unintended developer feature. I'd class it as a "nice to have" 😄

**Comment by jakebailey** [comment]
Definitely something I want to do (or push for); I just plugin-ified the eslint rules on the TS repo, so I think this is a logical step to address. Just have to work out the details of how dtslint runs eslint so that a plugin doesn't double up the plugin declaration (or vice versa).

**Pull request merged**